internal/http/tmdb/search.go

```diff
@@ -0,0 +1,323 @@
+// to whittle them down to a singular result. To do so, the year and popularity
+// of the results is taken in to consideration
+func (searcher *tmdbSearcher) handleSearchResults(results []SearchResultItem, metadata *media.FileMediaMetadata) (*SearchResultItem, error) {
+	if metadata.Year != nil {
+		if metadata.Episodic {
+			filterResultsInPlace(&results, metadata, func(resultDate time.Time, metadataDate time.Time) bool {
+				return resultDate.Compare(metadataDate) >= 0
+			})
+		} else {
+			filterResultsInPlace(&results, metadata, func(resultDate time.Time, metadataDate time.Time) bool {
+				return resultDate.Compare(metadataDate) == 0
+			})
+		}
+	}
+
+	if len(results) == 1 {
+		return &results[0], nil
+	} else if len(results) == 0 {
+		return nil, &NoResultError{}
+	}
+
+	metric := &metrics.Hamming{CaseSensitive: false}
+	stringSimilarity := make([]float64, len(results))
+	for i, res := range results {
+		stringSimilarity[i] = strutil.Similarity(res.Title, metadata.Title, metric)
+	}
+
+	sort.SliceStable(results, func(i, j int) bool { return stringSimilarity[i] < stringSimilarity[j] })
+	if stringSimilarity[0] > stringSimilarity[1]+0.25 {
+		return &results[0], nil
+	}
+
+	return nil, &MultipleResultError{results}
+}
+
+func (entry *SearchResultItem) effectiveDate() *Date {
+	if entry.FirstAirDate != nil {
+		return entry.FirstAirDate
+	}
+
+	return entry.ReleaseDate
+}
+
+func (date *Date) UnmarshalJSON(dateBytes []byte) error {
+	trimmedDateString := string(dateBytes[1 : len(dateBytes)-1])
+	parsed, err := time.Parse(time.DateOnly, trimmedDateString)
+	if err != nil {
+		return fmt.Errorf("cannot unmarshal Date due to error: %s", err.Error())
+	}
+
+	*date = Date{parsed}
+	return nil
+}
+
+// filterResultsInPlace will filter the given array of results IN PLACE by modifying
+// the provided slice and returning
+func filterResultsInPlace(results *[]SearchResultItem, metadata *media.FileMediaMetadata, filterFn func(dateFromResult time.Time, dateFromMetadata time.Time) bool) {
+	timeFromYear := func(year int) time.Time {
+		return time.Date(year, 1, 1, 0, 0, 0, 0, time.UTC)
+	}
+
+	yearFromMetadata := timeFromYear(*metadata.Year)
+	insertionIndex := 0
+	for _, v := range *results {
+		yearFromResult := timeFromYear(v.effectiveDate().Year())
+		if filterFn(yearFromResult, yearFromMetadata) {
+			(*results)[insertionIndex] = v
+			insertionIndex++
+		}
+	}
+
+	*results = (*results)[:insertionIndex]
+}
+
+func httpGetJsonResponse(urlPath string, targetInterface interface{}) error {
+	fmt.Printf("GET -> %s\n", urlPath)
+	resp, err := http.Get(urlPath)
+	if err != nil {
+		return &UnknownRequestError{fmt.Sprintf("failed to perform GET(%s) to TMDB: %s", urlPath, err.Error())}
+	}
+
+	defer resp.Body.Close()
+	respBody, err := io.ReadAll(resp.Body)
+
+	if resp.StatusCode != http.StatusOK {
+		var tmdbError tmdbError
+		if err := json.Unmarshal(respBody, &tmdbError); err != nil {
```

Comment:
Prefer decoding directly from the stream to avoid allocation of a byte array for the response body.

```suggestion
		if err := json.NewDecoder(resp.Body).Decode(&tmdbError); err != nil {
```

Reply:
Note that you will also need to remove line 275 if you do it this way.

Reply:
Ooooh TIL, thank you!